File: refine_domain_seq.pl
#! /usr/bin/perl
use strict;
use Bio::Seq;
use Bio::SeqIO;
#use Math::BigFloat;

die "perl $0 <hmmout> <protein.fasta> <outfile> <e-value>" unless (@ARGV == 4);

my $hmmout = Bio::SeqIO -> new (-file => "$ARGV[1]", 
	                            -format => "Fasta");

my $domfasta = Bio::SeqIO -> new (-file => ">$ARGV[2]", 
	                            -format => "Fasta");

my %keep=();
open IN, "$ARGV[0]" or die "$!";
while (<IN>){
	chomp;
	next if /^#/;

	my @arr = split(/\s+/,$_);
	next if $arr[6] > $ARGV[3];
	my @brr = ($arr[17],$arr[18]);
	my $target = $arr[0];
	if(!exists $keep{$target}){
		$keep{$target} = \@brr;
	}
}
close (IN);

while (my $seq = $hmmout -> next_seq()){
	my ($id,$sequence,$desc) = ($seq -> id,$seq -> seq,$seq -> desc);
	if(exists $keep{$id}){
		print "$id $keep{$id}[0] $keep{$id}[1]\n";
		### use trunc
        my $newseqobj = $seq -> trunc($keep{$id}[0],$keep{$id}[1]);
        ### use subseq
        # my $newseq = $seq -> subseq($keep{$id}[0],$keep{$id}[1]);

        # my $newseqobj = Bio::Seq ->new(-seq => $newseq,
        # 	                          -desc => 'subseq',
        # 	                          -id => $id);
		$domfasta -> write_seq($newseqobj);
	}
}
$hmmout -> close();
$domfasta -> close();

File: refine_domain_seq.py
# -*- coding: UTF-8 -*-
# FileName  : refine_domain_seq.py
# Time      : 2024/7/3
# Author    : xian

import sys
from Bio import SeqIO

# 检查命令行参数的数量是否为4，否则打印用法并退出
if len(sys.argv) != 5:
    print("Usage: python script.py <hmmout> <protein.fasta> <outfile> <e-value>")
    sys.exit(1)

# 从命令行参数中获取输入文件和参数
hmmout_file = sys.argv[1]
protein_fasta = sys.argv[2]
outfile = sys.argv[3]
e_value_threshold = float(sys.argv[4])

# 解析蛋白质序列文件
hmmout = SeqIO.parse(protein_fasta, "fasta")
# 准备输出文件，以FASTA格式写入
domfasta = open(outfile, "w")

# 初始化一个字典，用于保存满足条件的序列
keep = {}

# 打开hmmout文件并读取
with open(hmmout_file) as f:
    for line in f:
        line = line.strip()
        # 跳过以#开头的注释行
        if line.startswith("#"):
            continue
        
        # 按空白字符分割行内容
        arr = line.split()
        # 如果e-value大于阈值，跳过该行
        if float(arr[6]) > e_value_threshold:
            continue
        
        target = arr[0]
        brr = (int(arr[17]), int(arr[18]))
        
        # 如果target不在字典中，则添加
        if target not in keep:
            keep[target] = brr

# 重新打开蛋白质序列文件，处理每个序列
with open(protein_fasta) as f:
    for seq_record in SeqIO.parse(f, "fasta"):
        id = seq_record.id
        # 如果序列ID在字典中，处理该序列
        if id in keep:
            start, end = keep[id]
            print(f"{id} {start} {end}")
            
            # 提取子序列，BioPython使用0-based索引
            new_seq = seq_record[start-1:end]
            new_seq.id = seq_record.id
            new_seq.description = seq_record.description
            
            # 将新的子序列写入输出文件
            SeqIO.write(new_seq, domfasta, "fasta")

# 关闭输出文件
domfasta.close()